### ISSUE TITLE
Regarding 'sigaction'

### ISSUE THREAD

**Issue by pjhenning** [emscripten-core/emsdk#453]
Hello, when looking over the output of running `emconfigure ./configure` for a project I am trying to port to WASM, I noticed that the configure script decides that `sigaction()` is available (the relevant line of output is "checking for sigaction... yes"). The problem with this is that from what I can gather it seems `sigaction()` is not supported with emscripten (see: https://github.com/emscripten-core/emscripten/issues?q=is%3Aissue+sigaction). As a result, I am encountering similar errors to those reported in many of the issues linked above, namely, errors corresponding with the message: `Calling stub instead of sigaction()`.

In my opinion, emconfigure should ensure that configure scripts do not mistakenly conclude that sigaction is available, and that not doing so is a bug. Can anyone recommend steps to mitigate this issue, or perhaps correct me if I am mistaken about any of this?

**Comment by kripken** [comment]
I think your analysis is right. The other side though is that those stubs are useful when porting projects, as sometimes a project will only build if it finds sigaction or something else is present, and in practice it's fine if it doesn't do anything when called. And the errors ("Calling stub") are only shown in a build with assertions enabled, so they are only mean to help debugging.

However, perhaps in strict mode we should behave more strictly and not report such functions as present? I'm not sure what's best overall.

Btw, which project did you notice that on?

**Comment by sbc100** [comment]
I agree, perhaps we should have some kind of mode to disable all the stubs functions and then activate this mode when we detect autoconf or the the cmake-equivalent.

**Comment by pjhenning** [comment]
Yes at the very least some kind of opt-in mode which does not report such functions as available would be extremely useful.

The project I am having this issue with an attempt to port Mercury (https://github.com/Mercury-Language/mercury) code to WASM, which requires as a prerequisite that the Mercury runtime and standard library be built to WASM. I believe the call to sigaction is being made by the garbage collection system which is actually a separate codebase (https://github.com/ivmai/bdwgc), so I am trying to find out if there is a way to configure the garbage collector to avoid sigaction (https://github.com/ivmai/bdwgc/issues/312).

**Comment by sbc100** [comment]
As a workaround. you can usually pre-configure certain result via CONFIG_SITE which should cause autoconf to skip the test and instead accept the provided values.

**Comment by pjhenning** [comment]
Gonna give this a shot; can anyone point me to where the emconfigure functionality that tells configure scripts where to find the libc headers is?

**Comment by pjhenning** [comment]
I'm thinking that there can be two different versions of signal.h, one with declarations for the stub functions and one without, and emconfigure can conditionally point to one or the other.

**Comment by pjhenning** [comment]
Never mind I found it, sorry about the noise.

**Comment by sbc100** [comment]
I would imagine this would be link time thing rather than a compile time thing.  So the headers shouldn't need to change.

You probably want to have the setting disable the stubs that live in `src/library_signals.js`.

I imagine we will can to call this setting something like `DISABLE_JS_STUBS` and have it effect any JS functions that are purely stubs (error/asserts/etc).     Then we will probably want to enable this settings by default when we detect autoconf/cmake is in use.

**Comment by pjhenning** [comment]
The attempt I made here: https://github.com/pjhenning/emscripten/commit/087691ef9fe5b5f57f618da97183cd83177ae24d#diff-e9945c7364d1109b1d48f2d45c287800 does not work, which corroborates your stance on headers not needing to change, however I am confused about how making a change to `library_signals.js` will influence autoconf scripts. Does `emconfigure` somehow use the functions defined in the js library files?

**Comment by sbc100** [comment]
Configure script attempt to compile and link small programs that just reference a symbol.   If the emcc link command succeeds the check will pass.   When emcc is run as a linker symbols from `library_signals.js` and other libraries are used.  If you remove a symbol from a js library, for example by putting it within an `ifdef` block, then the link will fail, and autoconf will assume the function is missing (which i think the desired effect here).

**Comment by pjhenning** [comment]
I see, thank you for explaining! Will give that a shot then 😎

**Comment by sbc100** [comment]
These days we have `-sALLOW_UNIMPLEMENTED_SYSCALLS=0` which will disable these stubs.
